projects/driver_controls/src/throttle.c

```diff
@@ -0,0 +1,160 @@
+// The reading_ok_flag basically holds the actual state of reading_updated_flag as it resets.
+// In this module, it is assumed that channels hold a linear relationship with respect to the
+// pedal's position. This is used to verify if the readings are valid and "real" i.e if channels
+// are synced. If the data turns out to be stale or channels aren't synced, a timeout event will
+// be raised. The throttle_get_position function simply reads the position from storage.
+#include "throttle.h"
+#include <string.h>
+#include "event_queue.h"
+#include "input_event.h"
+
+// Scales a reading to a measure out of 12 bits based on the given range(min to max).
+static uint16_t prv_scale_reading(int16_t reading, int16_t max, int16_t min) {
+  return (1 << 12) * (reading - min) / (max - min);
+}
+
+// Given a reading from main channel and a zone, finds how far within that zone the pedal is pushed.
+// The scale goes from 0(0%) to 2^12(100%). For ex. pedal is at 2^11(50%) in brake zone.
+static uint16_t prv_get_numerator_zone(int16_t reading_main, ThrottleZone zone,
+                                       ThrottleStorage *storage) {
+  ThrottleZoneThreshold *zone_thresholds_main = storage->calibration_data->zone_thresholds_main;
+  return prv_scale_reading(reading_main, zone_thresholds_main[zone].max,
+                           zone_thresholds_main[zone].min);
+}
+
+// Given a reading from main channel and a zone, finds if the reading belongs to that zone.
+static bool prv_reading_within_zone(int16_t reading_main, ThrottleZone zone,
+                                    ThrottleStorage *storage) {
+  ThrottleZoneThreshold *zone_thresholds_main = storage->calibration_data->zone_thresholds_main;
+  return ((reading_main <= zone_thresholds_main[zone].max) &&
+          (reading_main >= zone_thresholds_main[zone].min));
+}
+
+// Returns true if the channel readings follow their linear relationship.
+static bool prv_channels_synced(int16_t reading_main, int16_t reading_secondary,
+                                ThrottleStorage *storage) {
+  ThrottleZoneThreshold *zone_thresholds_main = storage->calibration_data->zone_thresholds_main;
+  // Check if the reading_main is out of bounds.
+  if ((reading_main < zone_thresholds_main[THROTTLE_ZONE_BRAKE].min) ||
+      (reading_main > zone_thresholds_main[THROTTLE_ZONE_ACCEL].max)) {
+    return false;
+  }
+
+  ThrottleLine *line = storage->calibration_data->line;
+
+  int16_t max_main = line[THROTTLE_CHANNEL_MAIN].full_throttle_reading;
+  int16_t min_main = line[THROTTLE_CHANNEL_MAIN].full_brake_reading;
+
+  int16_t max_secondary = line[THROTTLE_CHANNEL_SECONDARY].full_throttle_reading;
+  int16_t min_secondary = line[THROTTLE_CHANNEL_SECONDARY].full_brake_reading;
+
+  int16_t tolerance = storage->calibration_data->tolerance;
+
+  // This condition checks for edge cases where readings are valid but not on the line.
+  if (reading_main < min_main) {
+    reading_main = min_main;
```

Comment:
You can use the `MIN`/`MAX` macros in `misc.h`. This would also mark values that are obviously out of range (0) correct.